pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+	"sort"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+type backoffInfo struct {
+	key                      string
+	failuresAfterLastSuccess int32
+	lastFailureTime          *metav1.Time
+}
+
+type backoffInfoStore struct {
+	store cache.Store
+}
+
+func (s *backoffInfoStore) getBackoffInfo(key string) *backoffInfo {
+	b, exists, _ := s.store.GetByKey(key)
+	if !exists {
+		return nil
+	}
+	return b.(*backoffInfo)
+}
+
+func (s *backoffInfoStore) updateBackoffInfo(key string, failuresAfterLastSuccess int32, lastFailureTime *metav1.Time) {
+	b, ok, _ := s.store.GetByKey(key)
+
+	if !ok {
+		s.store.Add(&backoffInfo{
+			key:                      key,
+			failuresAfterLastSuccess: failuresAfterLastSuccess,
+			lastFailureTime:          lastFailureTime,
+		})
+	} else {
+		backoffInfo := b.(*backoffInfo)
+		backoffInfo.failuresAfterLastSuccess = failuresAfterLastSuccess
+		backoffInfo.lastFailureTime = lastFailureTime
+	}
+}
+
+func (s *backoffInfoStore) removeBackoffInfo(jobId string) {
+	b, exists, _ := s.store.GetByKey(jobId)
+	if exists {
+		s.store.Delete(b)
+	}
+}
+
+func newBackoffInfoStore() *backoffInfoStore {
+	return &backoffInfoStore{
+		store: cache.NewStore(backoffInfoKeyFunc),
+	}
+}
+
+var backoffInfoKeyFunc = func(obj interface{}) (string, error) {
+	if u, ok := obj.(*backoffInfo); ok {
+		return u.key, nil
+	}
+	return "", fmt.Errorf("could not find key for obj %#v", obj)
+}
+
+func getNewBackoffInfo(backoffInfoStore *backoffInfoStore, key string, pods []*v1.Pod, newSucceededPods []*v1.Pod, newFailedPods []*v1.Pod) backoffInfo {
+	if backoffInfoStore.getBackoffInfo(key) == nil {
+		countedPods := getCountedPods(pods, newSucceededPods, newFailedPods)
+		failuresAfterLastSuccess, lastFailureTime := getInitialBackoffParameters(countedPods)
+		backoffInfoStore.updateBackoffInfo(key, failuresAfterLastSuccess, lastFailureTime)
+	}
+
+	backoffInf := backoffInfoStore.getBackoffInfo(key)
```

Comment:
Good catch on the double-count part. I have made the changes as you suggested. I removed the call to function `updateBackoffInfo`. But, I am not able to figure out how `getInitialBackoffParameters` can be eliminated. Do you mean copying the logic of `getInitialBackoffParameters` to this function? It will be great if you can throw some light on this.

Reply:
IIUC `getInitialBackoffParameters` computes the `failuresAfterLastSuccess, lastFailureTime` pair assuming no previous failures. But I think this is not necessary - we can just use zero values for the pair of initial values. The new values will be computed by the code of `getNewBackoff`.

Reply:
That is true when the job controller is running fine and a new job gets created. But, the purpose of `getInitialBackoffParameters` is to compute backoff when the job is already created and the job controller restarts due to some reason.

If the job is already in progress and the job controller restarts, the job controller looks back into the past on the pod success and failures and computes the backoff. That's what `getInitialBackoffParameters` does. If we remove, the backoff computation becomes non-resilient to controller restarts. Is that fine?

Reply:
Oh, now I see! I thought that this PR is just a preparatory refactoring to decouple the logic to make the resilience  to restart easier. The master code of the job controller also loses track of the backoff in case of restart so there is no regression. I thought there would be a follow up PR to support recomputing the backoff after restart. 

I guess we could fix the restart handling in this PR indeed. WDYT @alculquicondor ? 

But still, would it be possible to refactor the `getNewBackoff` to reuse its logic to compute the initial parameters after restart, I guess the logic would be the same, just the input pods are different (counted pods). Then I think we could compute the counted pods as non-active pods which are also not uncounted - to be sure we are not double counting.

Reply:
@mimowo , the changes are based on the suggestions that you provided [here](https://github.com/kubernetes/kubernetes/pull/114684#pullrequestreview-1230039644)

Reply:
I see, I just thought there will be two PRs - the preparatory decoupling here, and a follow up to fix the handling of restarts based on the counted pods, which seems cleaner. But, we could also combine the two steps into a single PR, I'm fine with both - asking Aldo for input. 

Anyway, I would like to make sure that we compute the counted pods in a way that makes it clear that they do not overlap with uncounted pods.

Reply:
Not sure I understand how you plan to recover the information after a controller restart. The desired behavior should be that once there is a succeeded finished pod, the backoff resets.

But in any case, dealing with controller restarts could be left for a follow up.

Reply:
Removed the handling of controller restarts.